pkg/cluster/cluster.go

```diff
@@ -0,0 +1,293 @@
+	if err != nil {
+		return nil, err
+	}
+
+	kconfig, err := clientcmd.BuildConfigFromFlags("", kubeConfigFilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	clientSet, err := kubernetes.NewForConfig(kconfig)
+	if err != nil {
+		return nil, err
+	}
+	return clientSet, nil
+}
+
+// GetCrdClientSet returns apiextclientset interface
+func GetCrdClientSet(clName string) (apiextclientset.Interface, error) {
+	kubeConfigFilePath, err := GetKubeConfigPath(clName)
+	if err != nil {
+		return nil, err
+	}
+
+	kconfig, err := clientcmd.BuildConfigFromFlags("", kubeConfigFilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	apiExtClientSet, err := apiextclientset.NewForConfig(kconfig)
+	if err != nil {
+		return nil, err
+	}
+	return apiExtClientSet, nil
+}
+
+// FinalizeSetup creates custom environment
+func FinalizeSetup(cl *Config, box *packr.Box, wg *sync.WaitGroup) error {
+	masterIP, err := GetMasterDockerIP(cl.Name)
+	if err != nil {
+		return err
+	}
+
+	err = PrepareKubeConfigs(cl.Name, cl.KubeConfigFilePath, masterIP)
+	if err != nil {
+		return err
+	}
+
+	clientSet, err := GetClientSet(cl.Name)
+	if err != nil {
+		return err
+	}
+
+	apiExtClientSet, err := GetCrdClientSet(cl.Name)
+	if err != nil {
+		return err
+	}
+
+	switch cl.Cni {
+	case "calico":
+		calicoDeploymentFile, err := GenerateCalicoDeploymentFile(cl, box)
+		if err != nil {
+			return err
+		}
+
+		calicoCrdFile, err := box.Resolve("tpl/calico-crd.yaml")
+		if err != nil {
+			return err
+		}
+
+		err = deploy.CrdResources(cl.Name, apiExtClientSet, calicoCrdFile.String())
+		if err != nil {
+			return err
+		}
+
+		err = deploy.Resources(cl.Name, clientSet, calicoDeploymentFile, "Calico")
+		if err != nil {
+			return err
+		}
+
+		err = wait.ForDaemonSetReady(cl.Name, clientSet, "kube-system", "calico-node")
+		if err != nil {
+			return err
+		}
+
+		err = wait.ForDeploymentReady(cl.Name, clientSet, "kube-system", "coredns")
+		if err != nil {
+			return err
+		}
+	case "flannel":
+		flannelDeploymentFile, err := GenerateFlannelDeploymentFile(cl, box)
+		if err != nil {
+			return err
+		}
+
+		err = deploy.Resources(cl.Name, clientSet, flannelDeploymentFile, "Flannel")
+		if err != nil {
+			return err
+		}
+
+		err = wait.ForDaemonSetReady(cl.Name, clientSet, "kube-system", "kube-flannel-ds-amd64")
+		if err != nil {
+			return err
+		}
+
+		err = wait.ForDeploymentReady(cl.Name, clientSet, "kube-system", "coredns")
+		if err != nil {
+			return err
+		}
+	case "weave":
+		weaveDeploymentFile, err := GenerateWeaveDeploymentFile(cl, box)
+		if err != nil {
+			return err
+		}
+
+		err = deploy.Resources(cl.Name, clientSet, weaveDeploymentFile, "Weave")
+		if err != nil {
+			return err
+		}
+
+		err = wait.ForDaemonSetReady(cl.Name, clientSet, "kube-system", "weave-net")
+		if err != nil {
+			return err
+		}
+
+		err = wait.ForDeploymentReady(cl.Name, clientSet, "kube-system", "coredns")
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
The contents of each switch clause should be moved into functions that are related to cluster creation but each CNI in a different file. You could even do it in a `map[string]function`, remove the switch and allow for registration of CNIs into that `map`. This, together with having a single `--cni` flag instead of multiple would allow for easy extending of support for other CNIs